core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/api/split.kt

```diff
@@ -46,22 +50,52 @@ internal fun <T, C, R> splitImpl(
                 columnCollectors[j].add(clause.default)
         }
 
-        var names = columnNamesGenerator(column, columnCollectors.size)
-        if (names.size < columnCollectors.size) {
-            names = names + (1..(columnCollectors.size - names.size)).map { "split$it" }
-        }
+        val names = columnNamesGenerator(column, columnCollectors.size)
+        val sourcePath = node.pathFromRoot()
+
+        columnCollectors.forEachIndexed { i, col ->
+            val preferredName = names.getOrNull(i)
 
-        columnCollectors.mapIndexed { i, col ->
+            val pathToInsert = if (clause.inward) sourcePath else sourcePath.dropLast(1)
+            val name = generateUnusedName(removeResult.df, preferredName, pathToInsert, columnsToInsert)
+
+            val path = pathToInsert + name
 
-            val name = nameGenerator.addUnique(names[i])
-            val sourcePath = node.pathFromRoot()
-            val path = if (clause.inward) sourcePath + name else sourcePath.dropLast(1) + name
             val data = col.toColumn(name)
-            ColumnToInsert(path, data, node)
+            columnsToInsert.add(ColumnToInsert(path, data, node))
+        }
+    }
+
+    return removeResult.df.insertImpl(columnsToInsert)
+}
+
+internal fun generateUnusedName(
+    df: DataFrame<*>,
+    preferredName: String?,
+    insertPath: ColumnPath,
+    columnsToBeInserted: List<ColumnToInsert>
+): String {
+    // check if column with this name already exists in the df in the same position in the hierarchy,
+    // or we already have a column with this name in the list of columns to be inserted to the same position in the hierarchy
+    fun isUsed(name: String) =
```

Comment:
Not very fond of inner functions, but you do reuse arguments from the outer function and it makes things more readable, so I think I'll let it slide :)